subprojects/core/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheEntryRetention.java

```diff
@@ -0,0 +1,75 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.caching.local.internal;
+
+import org.gradle.api.internal.cache.CacheConfigurationsInternal;
+import org.gradle.api.internal.cache.CacheResourceConfigurationInternal;
+import org.gradle.caching.local.DirectoryBuildCache;
+import org.gradle.internal.time.TimeFormatting;
+import org.gradle.internal.time.TimestampSuppliers;
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class DirectoryBuildCacheEntryRetention {
+    private final Supplier<Long> entryRetentionTimestampSupplier;
+    private final String description;
+
+    public DirectoryBuildCacheEntryRetention(DirectoryBuildCache directoryBuildCache, CacheConfigurationsInternal cacheConfigurations) {
+        @SuppressWarnings("deprecation")
+        int deprecatedRemoveUnusedEntriesAfter = directoryBuildCache.getRemoveUnusedEntriesAfterDays();
+
+        // Use the deprecated retention period if configured on `DirectoryBuildCache`, or use the core 'buildCache' cleanup configuration if not.
+        // If the deprecated property remains at the default, we can safely use the central value (which has the same default).
+        if (deprecatedRemoveUnusedEntriesAfter != CacheConfigurationsInternal.DEFAULT_MAX_AGE_IN_DAYS_FOR_BUILD_CACHE_ENTRIES) {
+            this.entryRetentionTimestampSupplier = TimestampSuppliers.daysAgo(deprecatedRemoveUnusedEntriesAfter);
+            this.description = "after " + deprecatedRemoveUnusedEntriesAfter + " days";
+            return;
+        }
+
+        CacheResourceConfigurationInternal buildCacheConfig = cacheConfigurations.getBuildCache();
+        this.entryRetentionTimestampSupplier = buildCacheConfig.getEntryRetentionTimestampSupplier();
+        this.description = describeEntryRetention(buildCacheConfig.getEntryRetention().get());
+    }
+
+    public Supplier<Long> getEntryRetentionTimestampSupplier() {
+        return entryRetentionTimestampSupplier;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    private static String describeEntryRetention(CacheResourceConfigurationInternal.EntryRetention entryRetention) {
+        long entryRetentionMillis = entryRetention.getTimeInMillis();
+
+        if (entryRetention.isRelative()) {
+            long expiryDays = TimeUnit.MILLISECONDS.toDays(entryRetentionMillis);
+            if (expiryDays >= 2) {
+                return "after " + expiryDays + " days";
+            } else {
+                return "after " + TimeFormatting.formatDurationTerse(entryRetentionMillis);
+            }
+        }
+
+        // Always render the timestamp in UTC
+        return "at " + Instant.ofEpochMilli(entryRetentionMillis).atZone(ZoneId.of("UTC")).format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss z"));
```

Comment:
🤔 Do we do some time rendering somewhere else? What pattern do we use there?

Reply:
Yes, but nothing consistent. For Build Scans it's important that we include the TZ because the date could be viewed in a different location to where the build executes.